html/beps/bep_0046.rst

```diff
@@ -0,0 +1,238 @@
+   downloads would be a lot higher. Each HAVE message is 9 bytes (length prefix,
+   type and piece index) and possibly another 20 bytes if sent on an otherwise
+	idle connection (TCP/IP overhead). Compared to the same torrent with a piece
+   size common today (2 MiB), it would require 128 times as many messages.
+
+Issue (1) is addressed by merkle tree torrents, in BEP30_.
+
+.. _BEP30: http://bittorrent.org/beps/bep_0030.html
+
+This BEP attempts to address (2) and (3) with a few different extensions.
+
+Redundant HAVEs
+===============
+
+A downloader that just completed a piece will send a HAVE message to all of
+its peers, announcing that it has a new piece. It does not take into account
+whether its peers also have this piece. The main reason to advertise
+pieces you have is to allow your peers to request them, making it unnecessary
+to tell peers that already have the piece. Presumably they are not interested
+in downloading it.
+
+A *redundant HAVE* refers to a HAVE message for a piece the
+receiving peer already has.
+
+There are some reasons to send redundant HAVEs:
+
+1. The uploader may want to track its upload progress to a peer, and get first
+   -hand feedback.
+2. Some super seed (BEP16_) implementations may rely on receiving HAVEs (for
+   example `libtorrent's`_).
+3. A client may want to be able to make rough download rate estimates of its
+   peers by measuring the timing of HAVE messages
+
+.. _BEP16: http://bittorrent.org/beps/bep_0016.html
+.. _`libtorrent's`: https://github.com/arvidn/libtorrent/commit/66ed31dd4b0c0b6d42c3eeb706477b1b19c8f1ea
+
+In order to not break existing clients, this BEP proposes a flag in the extended
+handshake (BEP10_) to let a peer indicate whether or not it need to receive
+redundant HAVEs.
+
+.. _BEP10: http://bittorrent.org/beps/bep_0010.html
+
+If the key ``rh`` exists and has a value of 0 in the extension handshake, the
+peer indicates that it does *not* need to receive redundant HAVE messages.
+
+If the ``rh`` key exists and is set to 1, the peer explicitly indicates that it
+is interested in redundant HAVE messages.
+
+.. parsed-literal::
+
+	{
+		"m": *extension messages*,
+		**"rh": 0**
+	}
+
+Clients implementing this extension are encouraged to always include this flag
+in their extension handshake, whether it's 0 or 1. This will enable a future
+transition of the default assumption, to be that peers do *not* need redundant
+HAVEs (unless explicitly stated).
+
+All have messages can be considered redundant to a peer that has declared
+itself to be *upload-only* (BEP21_).
+
+A high quality implementation may want to remember HAVE messages that would not
+have been redundant had the peer not been upload-only, and send them if the peer
+changes state to no longer be upload-only. Such bulk update could be compact
+with the extension message proposed in the following section.
+
+.. _BEP21: http://bittorrent.org/beps/bep_0021.html
+
+Compressed bitfields
```

Comment:
one theoretical upside is that it can reach beyond 2<sup>32</sup> pieces, but to actually make use of that we also would have to augment other piece-index messages